### PR TITLE
[IFP, ODB, DPL] Hybrid-cells placement implementation 

### PR DESCRIPTION
**Pull request by mina1460**

### CODE REVIEW

**Comment by maliberty** [comment]
Please fix the clang-tidy issues

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -266,7 +266,9 @@ add_subdirectory(cts)
 add_subdirectory(grt)
 add_subdirectory(tap)
 add_subdirectory(mpl)
+if (NOT ${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
 add_subdirectory(mpl2)
+endif()
```
Comment (maliberty) [inline src/CMakeLists.txt:271]: This is unrelated

File: src/odb/include/odb/db.h [hunk 2]
```diff
@@ -5237,6 +5237,39 @@ class dbSite : public dbObject
...
+
+  dbSite* getParent();
+
+  ///
+  /// Returns true if the row pattern is not empty
+  ///
+  bool hasRowPattern() const;
+
+  ///
+  /// Returns true if this cell is a hybrid cell
```
Comment (maliberty) [inline src/odb/include/odb/db.h:5259]: It would be good to describe what hybrid cell is.

File: src/odb/include/odb/db.h [hunk 3]
```diff
@@ -5237,6 +5237,39 @@ class dbSite : public dbObject
...
+  ///
+  void setRowPattern(
+      const std::vector<std::pair<dbSite*, dbOrientType>> row_pattern);
+
+  ///
+  /// sets the parent of the site
+  ///
+  void setParent(std::string parent_name);
+
+  dbSite* getParent();
```
Comment (maliberty) [inline src/odb/include/odb/db.h:5251]: Define what a parent site is.

File: src/dpl/src/Grid.cpp [hunk 4]
```diff
@@ -55,30 +55,138 @@ using utl::DPL;
...
+            calculateHybridSitesRowCount(working_site->getParent()),
+            db_row->getSiteCount(),
+            site_idx_to_grid_idx.at(working_site->getId()),
+            working_site->getParent()
+                ->getRowPattern(),  // FIXME(mina1460): this is wrong! in the
+                                    // case of HybridAB, and HybridBA. each of A
+                                    // and B maps to both, but the
+                                    // hybrid_sites_mapper only record one of
+                                    // them, resulting in the wrong RowPattern
+                                    // here.
```
Comment (QuantamHD) [inline src/dpl/src/Grid.cpp:180]: Does this still need to be addressed?

File: src/odb/include/odb/lefin.h [hunk 5]
```diff
@@ -212,6 +217,15 @@ class lefin
...
   void lineNumber(int lineNo);
+  const std::unordered_map<std::string, dbOrientType::Value> orientationMap
+      = {{"N", dbOrientType::R0},
+         {"W", dbOrientType::R270},
+         {"S", dbOrientType::R180},
+         {"E", dbOrientType::R90},
+         {"FN", dbOrientType::MYR90},
+         {"FW", dbOrientType::MY},
+         {"FS", dbOrientType::MXR90},
+         {"FE", dbOrientType::MX}};
```
Comment (maliberty) [inline src/odb/include/odb/lefin.h:228]: W = R90
E = R270
FN = MY
FS = MX
FW = MX90
FE = MY90

File: src/odb/src/db/dbSite.h [hunk 6]
```diff
@@ -94,6 +141,7 @@ inline dbIStream& operator>>(dbIStream& stream, _dbSite& site)
   stream >> site._height;
   stream >> site._width;
   stream >> site._next_entry;
+  stream >> site._row_pattern;
```
Comment (maliberty) [inline src/odb/src/db/dbSite.h:144]: This is a schema change.  This needs to be conditional on the schema version which needs to be incremented.

File: src/odb/src/db/dbSite.h [hunk 7]
```diff
@@ -52,7 +53,50 @@ struct dbSiteFlags
...
-  uint _spare_bits : 25;
+  uint _is_hybrid : 1;
+  uint _spare_bits : 24;
+};
+
+class RowPattern
+{
+ private:
+  uint n;
+  std::vector<std::pair<std::string, dbOrientType>> pattern;
```
Comment (maliberty) [inline src/odb/src/db/dbSite.h:64]: Is there a reason to use a string here rather than a dbId<dbSite>?

File: src/odb/src/lefin/lefin.cpp [hunk 8]
```diff
@@ -1690,6 +1690,23 @@ void lefin::site(lefiSite* lefsite)
...
+  if (lefsite->hasRowPattern()) {
+    auto row_pattern = lefsite->getRowPatterns();
+    std::vector<std::pair<dbSite*, dbOrientType>> converted_row_pattern;
+    converted_row_pattern.reserve(row_pattern.size());
+    for (auto& row : row_pattern) {
+      dbOrientType orient = dbOrientType::R0;
+      auto it = orientationMap.find(row.second);
+      if (it != orientationMap.end()) {
+        orient = it->second;
+      }
```
Comment (maliberty) [inline src/odb/src/lefin/lefin.cpp:1703]: This should always be found so you can just use at() to ensure it is.

File: src/odb/src/db/dbSite.h [hunk 9]
```diff
@@ -64,6 +108,8 @@ class _dbSite : public _dbObject
   uint _height;
   uint _width;
   dbId<_dbSite> _next_entry;
+  RowPattern _row_pattern;
+  std::string parent;
```
Comment (maliberty) [inline src/odb/src/db/dbSite.h:112]: Why a string?

**Review by maliberty — CHANGES_REQUESTED** [review]
feedback on the odb/ portion

File: src/ifp/include/ifp/InitFloorplan.hh [hunk 10]
```diff
@@ -120,6 +131,10 @@ class InitFloorplan
   dbBlock* block_;
   Logger* logger_;
   sta::dbNetwork* network_;
+
+ private:
```
Comment (maliberty) [inline src/ifp/include/ifp/InitFloorplan.hh:135]: Just change the whole protected section to private rather than make this one field private.  Nobody is inheriting from this class.

File: src/ifp/src/InitFloorplan.cc [hunk 11]
```diff
@@ -482,6 +499,122 @@ int InitFloorplan::makeRows(dbSite* site,
...
+          hybrid_site->getName(),
+          site_width,
+          site->getName(),
+          site->getWidth());
+    }
+  }
+
+  int core_height = abs(core_u.y() - core_l.y());
+  int core_width = abs(core_u.x() - core_l.x());
+  int rows_x = core_width / site_width;
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:534]: const

File: src/ifp/include/ifp/InitFloorplan.hh [hunk 12]
```diff
@@ -120,6 +131,10 @@ class InitFloorplan
...
+  // this is a set of sets of all constructed site ids.
+  std::set<std::set<int>> constructed_patterns;
```
Comment (maliberty) [inline src/ifp/include/ifp/InitFloorplan.hh:137]: class members should have a trailing _ in their name (see the others)

File: src/ifp/src/InitFloorplan.cc [hunk 13]
```diff
@@ -482,6 +499,122 @@ int InitFloorplan::makeRows(dbSite* site,
...
+  generateContiguousHybridRows(
+      parent_hybrid_site, row_pattern, patterns_to_construct);
+
+  for (const auto& pattern : patterns_to_construct) {
+    std::string pattern_sites = "[";
+    if (pattern.empty()) {
+      pattern_sites = "[]";
+    }
+    for (int i = 0; i < pattern.size(); i++) {
+      pattern_sites += pattern[i]->getName() + ",]"[i == pattern.size() - 1];
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:546]: That's an interesting trick

**Review by maliberty — CHANGES_REQUESTED** [review]
ifp feedback

File: src/dpl/include/dpl/Opendp.h [hunk 14]
```diff
@@ -145,13 +195,57 @@ struct Pixel
...
+  {
+  }
+
+  int getRowCount() const { return row_count; }
+
+  int getSiteCount() const { return site_count; }
+
+  int getGridIndex() const { return grid_index; }
+
+  const std::vector<std::pair<dbSite*, dbOrientType>>& getSites() const
```
[the remark this answers is not in the document]
Comment (maliberty) [reply src/dpl/include/dpl/Opendp.h:221]: That isn't true as a map isn't ordered and a vector is.

File: src/dpl/include/dpl/Opendp.h [hunk 15]
```diff
@@ -112,6 +112,32 @@ struct Master
   bool is_multi_row = false;
 };
 
+struct Grid_map_key
```
Comment (maliberty) [inline src/dpl/include/dpl/Opendp.h:115]: ```using GridMapKey = int;``` seems simpler

File: src/dpl/include/dpl/Opendp.h [hunk 16]
```diff
@@ -112,6 +112,32 @@ struct Master
...
+class HybridSiteInfo
+{
+ public:
+  HybridSiteInfo(int index, dbSite* site) : index(index), site(site) {}
+  int getIndex() { return index; }
+  dbSite* getSite() { return site; }
+
+ private:
+  int index;
+  dbSite* site;
```
Comment (maliberty) [inline src/dpl/include/dpl/Opendp.h:138]: Trailing _ on members

File: src/dpl/include/dpl/Opendp.h [hunk 17]
```diff
@@ -112,6 +112,32 @@ struct Master
...
+    return grid_index == other.grid_index;
+  }
+};
+
+class HybridSiteInfo
+{
+ public:
+  HybridSiteInfo(int index, dbSite* site) : index(index), site(site) {}
+  int getIndex() { return index; }
+  dbSite* getSite() { return site; }
```
Comment (maliberty) [inline src/dpl/include/dpl/Opendp.h:134]: const methods

File: src/dpl/include/dpl/Opendp.h [hunk 18]
```diff
@@ -478,8 +577,15 @@ class Opendp
...
 
   map<const dbMaster*, Master> db_master_map_;
-  map<int, GridInfo> grid_info_map_;
+  map<Grid_map_key, GridInfo> grid_info_map_;
+  unordered_map<int, int>
+      site_idx_to_grid_idx;  // this map is used to map each unqie site to a
+                             // grid. the key is always unique, but the value is
+                             // not unique in the case of hybrid sites
+                             // (alternating rows)
+  Grid_map_key smallest_non_hybrid_grid_key;
```
Comment (maliberty) [inline src/dpl/include/dpl/Opendp.h:586]: _ naming

File: src/dpl/include/dpl/Opendp.h [hunk 19]
```diff
@@ -478,8 +577,15 @@ class Opendp
...
   std::vector<GridInfo*> grid_info_vector_;
+  map<int, int> siteIdToGridId;
```
Comment (maliberty) [inline src/dpl/include/dpl/Opendp.h:588]: _ naming

File: src/dpl/src/Opendp.cpp [hunk 20]
```diff
@@ -476,13 +479,14 @@ int Opendp::gridPaddedWidth(const Cell* cell) const
...
 {
-  return divCeil(cell->height_, row_height);
+  return std::max(1, divCeil(cell->height_, row_height));
 }
 
 int Opendp::gridHeight(const Cell* cell) const
 {
   int row_height = getRowHeight(cell);
-  return divCeil(cell->height_, row_height);
+  // TODO: this is no longer correct for hybrid cells
```
Comment (maliberty) [inline src/dpl/src/Opendp.cpp:488]: ?? is this a concern

File: src/dpl/src/Opendp.cpp [hunk 21]
```diff
@@ -576,22 +583,50 @@ pair<int, GridInfo> Opendp::getRowInfo(const Cell* cell) const
...
+  }
+  auto site = cell->getSite();
+  if (!isStdCell(cell)) {
+    // non std cells can go to the first grid.
+    return smallest_non_hybrid_grid_key;
+  }
+  if (site == nullptr) {
+    logger_->error(DPL, 4219, "Cell {} has no site.", cell->name());
+  }
+  return this->getGridMapKey(site);
```
Comment (maliberty) [inline src/dpl/src/Opendp.cpp:624]: is 'this->' required?

File: src/dpl/test/report_failures.tcl [hunk 22]
```diff
@@ -1,9 +1,10 @@
 source "helpers.tcl"
 read_lef Nangate45/Nangate45.lef
 read_def report_failures.def
-catch { detailed_placement -report_file_name "report_failures.json" } msg
+set json_file [make_result_file report_failures.json]
+catch { detailed_placement -report_file_name $json_file } msg
```
Comment (maliberty) [inline src/dpl/test/report_failures.tcl:5]: nice fix

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/dpl/src/Grid.cpp (hunk 4 above)
Comment (maliberty) [reply src/dpl/src/Grid.cpp:192]: @mina1460 please respond

File: src/dpl/src/CheckPlacement.cpp [hunk 23]
```diff
@@ -65,9 +65,34 @@ void Opendp::checkPlacement(bool verbose,
   for (Cell& cell : cells_) {
     if (isStdCell(&cell)) {
       // Site alignment check
-      if (cell.x_ % site_width_ != 0 || cell.y_ % row_height_ != 0) {
-        site_align_failures.push_back(&cell);
+      if (!cell.isHybrid()) {
+        if (cell.x_ % site_width_ != 0 || cell.y_ % row_height_ != 0) {
+          site_align_failures.push_back(&cell);
+        }
+      } else {
```
[the remark this answers is not in the document]
Comment (maliberty) [reply src/dpl/src/CheckPlacement.cpp:72]: @QuantamHD I don't follow your comment as this is inside a loop.

**Comment by maliberty** [comment]
@mina1460 I see a few different open questions from @QuantamHD and myself.

File: src/dpl/src/Opendp.cpp [hunk 24]
```diff
@@ -389,6 +385,10 @@ bool Opendp::isStdCell(const Cell* cell) const
     case dbMasterType::ENDCAP_LEF58_LEFTBOTTOMCORNER:
     case dbMasterType::ENDCAP_LEF58_RIGHTTOPCORNER:
     case dbMasterType::ENDCAP_LEF58_LEFTTOPCORNER:
+      return true;
+    case dbMasterType::BLOCK:
+    case dbMasterType::BLOCK_BLACKBOX:
+    case dbMasterType::BLOCK_SOFT:
       // These classes are completely ignored by the placer.
```
Comment (QuantamHD) [inline src/dpl/src/Opendp.cpp:392]: You should move this comment above your changes.

Comment (maliberty) [reply src/dpl/src/Opendp.cpp:392]: Why?

Comment (QuantamHD) [reply src/dpl/src/Opendp.cpp:392]: Well they are now grouped with the COVER, RING and PAD classes. If we're doing that it seems like that comment would apply to them to. If it doesn't it almost feels like we should give them their own return value to differentiate it from the classes below.

Comment (maliberty) [reply src/dpl/src/Opendp.cpp:392]: Endcaps were always above that comment so nothing has changed in regards to it.  I just moved them to the 'return true' section.  They are not "grouped with the COVER, RING and PAD classes"

Comment (QuantamHD) [reply src/dpl/src/Opendp.cpp:392]: Seems fine then.

**Review by maliberty — APPROVED** [review]
More work remains on QOR but it is good enough to merge
